crates/red_knot_python_semantic/src/semantic_model.rs

```diff
@@ -58,8 +58,13 @@ impl HasTy for ast::ExpressionRef<'_> {
         let file_scope = index.expression_scope_id(*self);
         let scope = file_scope.to_scope_id(model.db, model.file);
 
-        let expression_id = self.scoped_ast_id(model.db, scope);
-        infer_scope_types(model.db, scope).expression_ty(expression_id)
+        if let Some(expression_id) = self.scoped_ast_id(model.db, scope) {
+            let lookup = infer_scope_types(model.db, scope).try_expression_ty(expression_id);
+            lookup.unwrap_or(Type::Unknown)
+        } else {
+            tracing::warn!("Couldn't find expression ID");
+            Type::Unknown
```

Comment:
I feel like it would make a lot of sense for there to be something like `Type::Unknown(SourcingInformation)`, where each callsite could embed where that specific one came from.

 That way, if `reveal_type` gives you an `Any`, you could probably trace back why you got that.